expression/constant_fold.go

```diff
@@ -23,12 +24,62 @@ func FoldConstant(expr Expression) Expression {
 	return e
 }
 
+func ifFoldHandler(expr *ScalarFunction) (Expression, bool) {
+	args := expr.GetArgs()
+	foldedArg0, _ := foldConstant(args[0])
+	if constArg, isConst := foldedArg0.(*Constant); isConst {
+		arg0, isNull0, err := constArg.EvalInt(expr.Function.getCtx(), nil)
+		if err != nil {
+			log.Warnf("fold constant %s: %s", expr.ExplainInfo(), err.Error())
+			return expr, false
+		}
+		if !isNull0 && arg0 != 0 {
+			return foldConstant(args[1])
+		}
+		return foldConstant(args[2])
+	}
+	isDeferredConst := false
+	var isDeferred bool
+	expr.GetArgs()[1], isDeferred = foldConstant(args[1])
+	isDeferredConst = isDeferredConst || isDeferred
+	expr.GetArgs()[2], isDeferred = foldConstant(args[2])
+	isDeferredConst = isDeferredConst || isDeferred
+	return expr, isDeferredConst
+}
+
+func ifNullFoldHandler(expr *ScalarFunction) (Expression, bool) {
+	args := expr.GetArgs()
+	foldedArg0, _ := foldConstant(args[0])
+	if constArg, isConst := foldedArg0.(*Constant); isConst {
+		_, isNull0, err := constArg.EvalInt(expr.Function.getCtx(), nil)
+		if err != nil {
+			log.Warnf("fold constant %s: %s", expr.ExplainInfo(), err.Error())
+			return expr, false
+		}
+		if isNull0 == true {
+			return foldConstant(args[1])
+		}
+	}
+	isDeferredConst := false
+	expr.GetArgs()[1], isDeferredConst = foldConstant(args[1])
+	return expr, isDeferredConst
+}
+
 func foldConstant(expr Expression) (Expression, bool) {
 	switch x := expr.(type) {
 	case *ScalarFunction:
 		if _, ok := unFoldableFunctions[x.FuncName.L]; ok {
 			return expr, false
 		}
+		// specialFoldHandler stores functions for special UDF to constant flod
+		var specialFoldHandler = map[string]func(*ScalarFunction) (Expression, bool){
```

Comment:
put this map outside this function.

Reply:
I have tried to put it in `expression/function_traits.go`. But it cause **Initialization Loop** error.

Reply:
Not just in `expression/function_traits.go`. In any place outside function will cause initialization loop.

Reply:
``` go
var specialFoldHandler = map[string]func(*ScalarFunction) (Expression, bool){}

func init() {
	// specialFoldHandler stores functions for special UDF to constant fold
	specialFoldHandler = map[string]func(*ScalarFunction) (Expression, bool){
		ast.If:     ifFoldHandler,
		ast.Ifnull: ifNullFoldHandler,
	}
}

```